lib/sorbet-coerce.rb

```diff
@@ -1,22 +1,34 @@
 # typed: strict
+require_relative 'configuration'
 require 'private/converter'
 require 'safe_type'
 
-module T
-  class CoercionError < SafeType::CoercionError; end
+module T::Coerce
+  class ValueError < SafeType::CoercionError; end
+  class ShapeError < SafeType::CoercionError; end
 
-  module Coerce
-    define_singleton_method(:[]) do |type|
-      Class.new(T::Private::Converter) do
-        define_method(:to_s) { "#{name}#[#{type.to_s}]" }
+  define_singleton_method(:[]) do |type|
+    Class.new(T::Private::Converter) do
+      define_method(:to_s) { "#{name}#[#{type.to_s}]" }
 
-        define_method(:from) do |args|
-          begin
-            T.send('let', send('_convert', args, type), type)
-          rescue TypeError
-            raise CoercionError.new(args, type)
-          end
+      define_method(:from) do |args, raise_value_error: nil|
+        @raise_value_error = T.let(raise_value_error, T.nilable(T::Boolean))
+        if raise_value_error.nil?
+          @raise_value_error = T::Coerce::Configuration.raise_value_error
         end
+
+        error = nil
+        begin
+          return T.send('let', send('_convert', args, type), type)
+        rescue TypeError
+          error = ValueError.new(args, type)
+        rescue ValueError => e
+          error = e
+        end
+
+        handler = T::Coerce::Configuration.value_error_handler
+        handler.nil? ? raise(error) : handler.call(error)
```

Comment:
The handler isn't very useful when it's here, because the process is already aborted. I was thinking about like people can handle the value case-by-case, ie accepting some and not accepting another, or supply a different value when there is an error. Anyway, for the purpose of moving this forward & fix the issue, we can discuss about a more sophisticated error handler later. Let's use the `raise_value_error` value as the main way to customize the handling for now.